roll.1
.\" Process this file with
.\" groff -man -Tascii foo.1
.\"
.TH roll 1 2021-08-23 roll "Roll User Manual"
.SH NAME
roll \- rolls a user-defined dice sequence
.SH SYNOPSIS
.BR "roll " "[" "-hpsv"  "] [" "--version" "] expression"
.SH DESCRIPTION
.B roll
Rolls a user-defined dice sequence. The dices to roll are defined
using dN where N is the number of sides. Dices can be rolled multiple
times by prepending the number of repetitions (e.g., 3d6) and used in
simple mathematical expressions (e.g., 2d8+4).
.SH OPTIONS
.TP
.B \-\-help, \-h
help
.TP
.B \-\-positive, \-p
allow only positive results
.TP
.B \-\-sum-series, \-s
show the sum of roll series
.TP
.B \-\-verbose, \-v
verbose (displays the result for each roll)
.TP
.B \--version
prints the version of the program
.SH EXPRESSIONS
Expressions are in dice notation (also known as dice algebra, common
dice notation or RPG dice notation).

Die rolls are expressed in the form NdX or NDX where N is the number of dices
(1 if omitted) and X the number of dice faces.

roll also accepts W (German) or T (Swedish) to specify a dice (D in English)

If the final number is omitted, it is assumed to be a six.

To this basic notation, an additive modifier can be appended, yielding
expressions of the form, NdX+C. The plus can be replaced by a minus
sign ("-") for subtraction. C is a number to be added subtracted from
the final total.

See: http://en.wikipedia.org/wiki/Dice_notation

.SH Percentile dice (d%)

Often, the variable X in the above notation will be "%". Although a
100-sided die does exist, it is both more common and more uniformly
random to use a combination of two ten-sided dice known as percentile
dice, where one die represents tens and the other units.  A roll of
two zeroes means 100.

If you want to roll a single 100-sided die use the "d100" notation.

.SH d10x

d10x is equivalent to d10 x d10 and gives a non-linear distribution.

.SH Fudge dice (dF or df)

The Fudge role-playing game system uses customized "Fudge dice" which
have an equal number of plus, minus and blank sides.

.SH Multipliers

In some games, the above notation is expanded to allow for a
multiplier, as in NdXxC or CxNdX, where "x" or "*" denotes
multiplication (can be replaced by "/" for division) and C is a
natural number.

Results are rounded up.

.SH Reroll filters

It is possible to specify restrictions on dice rolls telling which
values to keep and which values to discard by putting a filter after
the dice. If a value is to be discarded the dice is re-rolled until
the result is valid value. You can specify to keep only values bigger
(>), bigger or equal (>=), smaller (<), smaller or equal (<=) or
different (!=) from a given number. The comparator is specified right
after the the die specification (e.g., 1D6>2).

.SH Selective results

It is possible to keep only the M highest or lowest dices with the following notation:

.PD 0
.TP
NdXhM: roll N dices with X sides and keep the highest M results (M must be less or equal N)
.TP
NdXkM: K for keep is equivalent to H for highest
.TP
NdXlM: roll N dices with X sides and keep the lowest M results (M must be less or equal N)
.PD 0

.TP
Or using the abbreviated notation NdX-L to drop the lowest and NdX-H to drop the highest result.

.SH Series and repetitions

It is possible to concatenate top-level expressions with a comma
("expression, expression") and to repeat the execution of an
expression or a set of expressions by using a counter and curly
brackets ("N{ expression}")

.SH Examples

.PD 0
.TP
3d6        rolls 3 6-sided dices and sums the results
    (can be abbreviated with "3d")
.TP
1d8 + 3    rolls 1 8-sided die and adds 3
    (can be abbreviated by "d8+3")
.TP
d%         rolls 2 10-sided dices: one represents the tens and
    the other units
.TP
1d6x5      rolls 1 6-sided dice and multiplies the result by 5
.TP
1d6 / 2    rolls 1 6-sided dice and divides the result by 2 (1d3)
.TP
4d6h3      rolls 4 6-sided dices and keeps the 3 highest results
    (discarding the lowest)
.TP
4d6-L      rolls 4 6-sided dices and discards the lowest
.TP
"6{3d6}"   rolls 3d6 6 times (expression
    is quoted to avoid the > being interpreted as
    redirection by the shell)
.TP
"1d6>2"    rolls 1d6 until the result is bigger than 2 (expression
    is quoted to avoid the > being interpreted as
    redirection by the shell)
.TP
4DF        rolls 4 Fudge dices
.TP

.PD 1

.SH Grammar (EBNF)

.PD 0
.TP
expression   := term { "+" term | "-" term } .
.TP
term         := number | factor [ ("*"|"/") number |
         number ("*"|"/") factor | "(" expression ")" .
.TP
factor       := number FilteredDice [ ("h"|"H"|"l"|"L"|"k"|"K") number ] | number FilteredDice "-" [ ("h"|"H"|"l"|"L") | FilteredDice .
.TP
FilteredDice := dice | dice (">"|">="|"<"|"<="|"!=") number .
.TP
dice         := "d" | "D" | "w" | "W" | "t" | "T" [ number | "%" | "F" | "f" | "x" | "X" ] .
.PD 1

.SH "SEE ALSO"
.BR random (3)
.SH "EXIT STATUS"
roll returns a zero exist status if it finds no error in the
supplied expresssion
.SH BUGS
Please report bugs to: Matteo Corti (matteo (at) corti.li )

.SH AUTHOR
Matteo Corti ( matteo (at) corti.li )


NEWS
2021-08-24      2.6.0: fixed several bugs and updated the grammar
2019-04-18      2.5.0: uses urandom if possible
2018-07-15      2.4.0: Added the d10x notation
2018-07-04	2.3.0: Added the K keyword (keep)
2018-07-04      2.2.0: Added -L and -H notation (drop the lowest or highes dice)
2016-08-23	2.1.1: support for MinGW on Windows
2016-06-18	2.1.0: added 'W' and 'T' as dice for German and Swedish notation
2014-09-08      2.0.2: fixed an error in the initalization of parse tree
                       nodes
2011-08-27      2.0.0: added Fudge dice, reroll filters and series
                1.1.4: fixed an issue which prevented compilation on FreeBSD
2008-01-29      1.1.3: fixed a bug in the parsing of expressions with more
                       than two terms
2007-08-22      1.1.2: fix on the percentile dices
2007-03-18      1.1.0: percentile dices (d%), multipliers (10d6x2) and
                       selective results (see man page)
2007-03-12      1.0.4: correct exit codes and positive results only
2007-03-01	1.0.3: spec file for RPM generation
2007-02-26      1.0.2: small fix in the random number generator
                       initialization (see ChangeLog)


roll.c
/* Copyright (c) 2005-2019  Matteo Corti
 * This file is part of roll
 *
 * You may distribute this file under the terms the GNU General Public
 * License.  See the file COPYING for more information.
 */

/*!
 * \file   roll.c
 * \author Matteo Corti
 * \brief  The main application file
 */

#include "roll.h"
#include "parser.h"

int yyparse(void);

int sum_flag     = FALSE; /**< command line argument: sum series     */
static int test         = 0; /**< command line argument: test run (no random results) */
static int verbose_flag = FALSE; /**< command line argument: verbose output */
static int version_flag = FALSE; /**< command line argument: version        */

extern int positive_flag; /**< command line argument: allow only positive results */

/** Not defined in the automatically generated header files */
#ifndef YY_TYPEDEF_YY_BUFFER_STATE
#define YY_TYPEDEF_YY_BUFFER_STATE
typedef struct yy_buffer_state *YY_BUFFER_STATE;
#endif
YY_BUFFER_STATE yy_scan_string ( const char *yy_str  );

#ifdef DEBUG
int debug_flag = 0; /**< command line argument: debug output */
#endif

#ifdef DEBUG
void debug( char * message ) {
  if ( debug_flag > 0 ) {
    printf("[DBG] %s\n", message);
  }
}
#endif

/*!
 * \brief Prints the program's usage
 */
void usage() {
  
  printf("\nusage: %s [OPTION] expression\n\n", PACKAGE_NAME);
  printf("Options\n");
  printf(" -p, --positive     allow only positive results\n");
  printf(" -h, --help         show this help screen\n");
  printf(" -s, --sum-series   show the sum of roll series\n");
  printf(" -t, --test         the result of the dice is always 3 (for testing purposes)\n");
  printf(" -v, --verbose      increase verbosity\n");
  printf("     --version      prints the program version and exits\n");
#ifdef DEBUG
  printf(" -d, --debug        debugging output\n");
#endif
  printf("\nPlease see the %s(1) man page for full documentation\n\n", PACKAGE_NAME);

}

/*!
 * \brief          Prints the specified error message and exits with a failure status
 * \param[message] error message
 */
void error(char * message) {
  fprintf(stderr, "\nError: %s\n", message);
  exit(EXIT_FAILURE);
}

/*!
 * \brief        Wrapper for roll(int dice) translates special dices (e.g., d%)
 * \param[sides] Sides of the dice
 */
int roll_dice(int sides) {

  int result = 0;
  
  if ( sides == HUNDRED ) {
    
    /* d100 -> d10*10+d10 */
    
    int d10 = roll(10);
    if (verbose_flag) {
      printf("d10 -> %i\n", d10);
    }
    d10 = d10 % 10;
    
    int d1 = roll(10);
    if (verbose_flag) {
      printf("d10 -> %i\n", d1);
    }
    
    if (d1 == 0 && d10 == 0) {
      result = 100;
    } else if (d10 == 0) {
      result = d1;
    } else {
      result = d10*10 + d1;
    }

  } else if ( sides == FUDGE_DICE ) {

    result = roll(3) - 2;

    if (verbose_flag) {

      switch (result) {
      case -1:
        printf("dF -> -\n");
        break;
      case 0:
        printf("dF -> o\n");
        break;
      case 1:
        printf("dF -> +\n");
        break;        
      }
    }
    
  } else {

    result = roll(sides);
    
    if (verbose_flag) {
      printf("d%i -> %i\n", sides, result);
    }

  }

  return result;
  
}

/*!
 * \brief       Rolls an n-sided dice
 * \param[dice] Number of sides of the rolled dice
 * \return      Result of the dice roll
 */
int roll(int dice) {

  /* 
   * In: W. H. Press et al,Numerical Recipes in C: The Art of
   * Scientific Computing.  New York, Cambridge University Press,
   * 1992, 2nd ed., p. 277
   *
   * "If you want to generate a random integer between 1 
   *  and 10, you should always do it by using high-order
   *  bits, as in
   *
   *  j=1+(int) (10.0*rand()/(RAND_MAX+1.0));
   */

  int res;

  if ( ! test ) {
  
#if defined HAVE_SRANDOMDEV || defined HAVE_SRANDOM
    res = 1+(int)(((double)dice)*random()/(RAND_MAX+1.0));
#else
    res = 1+(int)(((double)dice)*rand()/(RAND_MAX+1.0));
#endif

  } else {
    res = test;
  }
  
  return res;

}

#ifdef HAVE_SRANDOM
/*!
 * \brief       Try to initialize random from /dev/urandom, otherwise fallback on srandom(time(0))
 */
static void urandom_init() {
  FILE* urandom = fopen("/dev/urandom", "rb");
  if (urandom) {
    unsigned int seed = 0;
    fread(&seed, sizeof(seed), 1, urandom);
    fclose(urandom);
    srandom(seed ^ time(NULL));
  }
  else
    srandom(time(NULL));
}
#endif

/*!
 * \brief       Main program
 * \param[argc] Number of command line arguments
 * \param[argv] Array of strings with the command line arguments
 * \return      Exit status
 */
int main(int argc, char **argv) {

  char   expression[EXPRESSION_SIZE];
  int    expression_size;

#ifdef DEBUG
  debug("[DBG] Debug mode enabled");
#endif

  
#ifdef HAVE_SRANDOMDEV
  srandomdev();
#elif defined HAVE_SRANDOM
  urandom_init();
#else
  srand(time(NULL));
#endif
     
  while (TRUE) {

    static struct option long_options[] = {
      {"sum-series",  no_argument,       NULL, 's'},
      {"positive",    no_argument,       NULL, 'p'},
      {"test",        no_argument,       NULL, 't'},
      {"verbose",     no_argument,       NULL, 'v'},
      {"version",     no_argument,       &version_flag, TRUE},
      {"help",        no_argument,       NULL, 'h'},
#ifdef DEBUG
      {"debug",       no_argument,       NULL, 'd'},
#endif
      {NULL, 0, NULL, 0}
    };

    /* getopt_long stores the option index here. */
    int option_index = 0;

    int c;
    
#ifdef DEBUG
    c = getopt_long (argc, argv, "hvsptd",
		     long_options, &option_index);
#else
    c = getopt_long (argc, argv, "hvspt",
		     long_options, &option_index);
#endif
    
    /* Detect the end of the options. */
    if (c == -1)
      break;
     
    switch (c) {

    case 'v':
      verbose_flag = TRUE;
      break;

    case 's':
      sum_flag = TRUE;
      break;
      
    case 'p':
      positive_flag = TRUE;
      break;

    case 'h':
      usage();
      exit(0);

    case 't':
      test = 3; /* value to be used as a fix result of a dice roll */
      break;

    case '?':
      usage();
      /* getopt_long already printed an error message. */
      exit(EXIT_SUCCESS);

#ifdef DEBUG
    case 'd':
      debug_flag++;
      break;      
#endif
      
    case 0:
      break;

    default:
      abort ();
    }

  }

  if (version_flag) {
    printf("%s %s\n", PACKAGE_NAME, PACKAGE_VERSION);
    exit(EXIT_SUCCESS);
  }      

  argc -= optind;
  argv += optind;         
  
  /* build string to parse */
  expression[0] = '\0';
  expression_size = 0;
  while(argc>0) {
    expression_size += strlen(*argv);
    if (expression_size >= EXPRESSION_SIZE) {
      error("Expression too long!\n");
    }
    strncat(expression, *argv, EXPRESSION_SIZE-1);
    argc--;
    argv++;
  }
  
  if (expression_size > 0) {

#ifdef DEBUG
    debug(expression);
#endif
    
    yy_scan_string(expression);

#ifdef DEBUG
    debug("expression scanned");
#endif
    
    yyparse();

#ifdef DEBUG
    debug("expression parsed\n");
#endif

  } else {
    error("No expression provided!\nPlease use the \"-h\" option.\n");
    exit(EXIT_FAILURE);
  }

  return 0;

}

/*!
 * \brief  Allocates a new IR node
 * \return Newly allocated node
 */
struct ir_node * allocate_node ( void  ) {

  struct ir_node * node = malloc(sizeof(struct ir_node));
  if (node == NULL) {
    error("Out of memory\n");
    exit(EXIT_FAILURE);
  }

  /* initialize default values */
  node->left  = NULL;
  node->right = NULL;
  node->next  = NULL;
  node->op    = 0;
  node->value = 0;
  
  return node;
  
}

/*!
 * \brief         Allocates a new NUMBER node
 * \param[number] Number value
 * \return        A NUMBER node
 */
struct ir_node * new_number ( int number ) {

  struct ir_node * node = allocate_node();
  node->op    = OP_NUMBER;
  node->value = number;

  return node;

}

/*!
 * \brief     Comparison function for qsort
 * \param[p1] First integer
 * \param[p2] Second integer
 * \return    1 if p1 > p2, -1 if p1 < p2 and 0 if p1 and p2 are equal
 */
int compare(const void * p1, const void * p2) {

  const int i1 = *((const int *)p1);
  const int i2 = *((const int *)p2);

  if (i1 > i2) {
    return 1;
  } else if (i1 < i2) {
    return -1;
  } else {
    return 0;
  }
  
}

/*!
 * \brief        Allocates a new OP node
 * \param[op]    Operation
 * \param[left]  Left operand
 * \param[right] Right operand
 * \return       An OP node
 */
struct ir_node * new_op ( unsigned short int op, struct ir_node * left, struct ir_node * right) {

  struct ir_node * node = allocate_node();
  node->op    = op;
  node->value = 0;
  node->left  = left;
  node->right = right;
  return node;
  
}

/*!
 * \brief        Allocates a new DICE node
 * \param[sides] A node representing the number of dice sides
 * \return       A DICE node
 */
struct ir_node * new_dice ( struct ir_node * sides) {
  
  struct ir_node * node = allocate_node();
  node->op    = OP_DICE;
  node->value = 0;
  node->right = sides;
  return node;
  
}

int checked_sum( int op1, int op2 ) {
  if ( (op2 > 0 && op1 > INT_MAX - op2) ||
       (op2 < 0 && op1 < INT_MIN - op2)) {
    error("Overflow");
  }
  return op1+op2;
}

int checked_multiplication( int op1, int op2 ) {
  int result = op1 * op2;
  if (op1 != 0 && result / op1 != op2 ) {
    error("Overflow");
  }
  return result;
}

/*!
 * \brief        Roll dices and compute expressions
 * \param[node]  The root of the expression tree to compute
 * \param[print] Set to true to print the result of the expression
 * \return       The expression value
 */
int roll_expression ( struct ir_node * node, int print ) {

  int  high;
  int  i;
  int  limit;
  int  low;
  int  repetitions;
  int  return_value = 0;
  int  sides;
  int  tmp;
  int * results;

  struct ir_node * cur;

  cur = node;
  while (cur != NULL) {

    int sum = 0;

    switch (cur->op) {
    
    case OP_NUMBER:
      sum = cur->value;
      break;

    case OP_REP:

      for (i = 0; i < roll_expression(cur->left, FALSE); i++) {
        sum = checked_sum( sum, roll_expression(cur->right, FALSE) );
      }
      break;
      
    case OP_DICE:
      sum = roll_dice( roll_expression(cur->right, FALSE) );
      break;
      
    case OP_PLUS:
      sum = checked_sum( roll_expression( cur->left,  FALSE ),
                         roll_expression( cur->right, FALSE ) );
      break;
      
    case OP_MINUS:
      sum = checked_sum( roll_expression( cur->left,  FALSE ),
                         -roll_expression( cur->right, FALSE ) );
      break;
      
    case OP_TIMES:
      sum = checked_multiplication( roll_expression( cur->left,  FALSE ),
                                   roll_expression( cur->right, FALSE ) );
      break;
      
    case OP_DIV:
      sum = (int)
        ceil( (float)roll_expression( cur->left,  FALSE ) /
              roll_expression( cur->right, FALSE ) );
      break;
      
    case OP_HIGH:

      sides       = roll_expression(cur->right->right->right, FALSE);
      repetitions = roll_expression(cur->right->left,  FALSE);
      high        = roll_expression(cur->left, FALSE);      

      /* array to store the results to sort */
      if (!(results = malloc(sizeof(int)*repetitions))) {
        error("Out of memory");
      }
      
      for(i=0; i<repetitions; i++) {
        results[i] = roll_dice(sides);
      }
      qsort(results, repetitions, sizeof(int), &compare);

      for(i=(repetitions-high); i<repetitions; i++) {
        sum = checked_sum( sum, results[i] );
      }
      
      free(results);
      
      break;
      
    case OP_LOW:
      
      sides       = roll_expression(cur->right->right->right, FALSE);
      repetitions = roll_expression(cur->right->left,  FALSE);
      low         = roll_expression(cur->left, FALSE);
      
      if (cur->right->left != NULL) {
        repetitions = roll_expression(cur->right->left, FALSE);
      }
                  
      /* array to store the results to sort */
      if (!(results = malloc(sizeof(int)*repetitions))) {
        error("Out of memory");
      }
      
      for(i=0; i<repetitions; i++) {
        results[i] = roll_dice(sides);
      }
      qsort(results, repetitions, sizeof(int), &compare);
      for(i=0; i<low; i++) {
        sum = checked_sum( sum, results[i] );
      }
      
      free(results);
      
      break;

    case OP_GT:
      
      limit = roll_expression(cur->right, FALSE);      
      tmp   = roll_expression(cur->left,  FALSE);
      while (tmp <= limit) {
        tmp = roll_expression(cur->left, FALSE);
      }
      sum = checked_sum( sum, tmp );
      
      break;
      
    case OP_GE:
      
      limit = roll_expression(cur->right, FALSE);      
      tmp   = roll_expression(cur->left,  FALSE);
      while (tmp < limit) {
        tmp = roll_expression(cur->left, FALSE);
      }
      sum = checked_sum( sum, tmp );
      
      break;
      
    case OP_LT:
      
      limit = roll_expression(cur->right, FALSE);      
      tmp   = roll_expression(cur->left,  FALSE);
      while (tmp >= limit) {
        tmp = roll_expression(cur->left, FALSE);
      }
      sum = checked_sum( sum, tmp );
      
      break;
      
    case OP_LE:
      
      limit = roll_expression(cur->right, FALSE);      
      tmp   = roll_expression(cur->left,  FALSE);
      while (tmp > limit) {
        tmp = roll_expression(cur->left, FALSE);
      }
      sum = checked_sum( sum, tmp );
      
      break;
      
    case OP_NE:
      
      limit = roll_expression(cur->right, FALSE);      
      tmp   = roll_expression(cur->left,  FALSE);
      while (tmp == limit) {
        tmp = roll_expression(cur->left, FALSE);
      }
      sum = checked_sum( sum, tmp );
      
      break;
      
    default :
      
      fprintf(stderr, "Implementation error: unkown IR node with code %i\n", cur->op);
      exit(EXIT_FAILURE);
      
    }

    return_value = checked_sum( return_value, sum);
    if (print == TRUE) {
      printf("%i\n", sum);
    }
    
    cur = cur->next;
    
  }

  return return_value;
  
}

#ifdef DEBUG

void print_node( struct ir_node * node) {

  switch (node->op) {
  case OP_NUMBER: printf("number (%i)", node->value); break;      
  case OP_DICE:   printf("dice"); break;
  case OP_PLUS:   printf("+"); break;
  case OP_MINUS:  printf("-"); break;
  case OP_TIMES:  printf("*"); break;
  case OP_DIV:    printf("/"); break;
  case OP_HIGH:   printf("high"); break;
  case OP_LOW:    printf("low"); break;
  case OP_GT:     printf(">"); break;
  case OP_GE:     printf(">="); break;
  case OP_LT:     printf("<"); break;
  case OP_LE:     printf("<="); break;
  case OP_NE:     printf("!="); break;
  case OP_REP:    printf("rep"); break;
  default :       printf("unknown node"); break;
  }
  
}

void print_tree( char * prefix, struct ir_node * node, int indent) {

  int i;
  
  printf("[%s] ", prefix);

  for (i = 0; i < indent; i++) {
    printf("  ");
  }

  print_node(node);

  printf("\n");

  if (node->left != NULL) {
    print_tree(prefix, node->left, indent+1);
  }

  if (node->right != NULL) {
    print_tree(prefix, node->right, indent+1);
  }
  
}
#endif
